Projects/Tetris/Program.cs

```diff
@@ -0,0 +1,768 @@
+                break;
+            case ConsoleKey.Z:
+                if (!DEBUGCONTROLS) return;
+                TETROMINO.Shape = TETROMINOS[6];
+                break;
+            case ConsoleKey.X:
+                if (!DEBUGCONTROLS) return;
+                Score += 10;
+                break;
+        }
+    }
+}
+
+void DrawFrame()
+{
+    bool collision = false;
+    int yScope = TETROMINO.Y;
+    string[] shapeScope = TETROMINO.Shape;
+    char[][] frame = new char[PLAYFIELD.Length][];
+
+    //Field
+    for (int y = 0; y < PLAYFIELD.Length; y++)
+    {
+        frame[y] = PLAYFIELD[y].ToCharArray();
+    }
+
+    //Draw Tetromino
+    for (int y = 0; y < shapeScope.Length && !collision; y++)
+    {
+        for (int x = 0; x < shapeScope[y].Length; x++)
+        {
+            int tY = yScope + y;
+            int tX = TETROMINO.X + x;
+            char charToReplace = PLAYFIELD[tY][tX];
+            char charTetromino = shapeScope[y][x];
+
+            if (charTetromino == ' ') continue;
+
+            if (charToReplace != ' ')
+            {
+                collision = true;
+                break;
+            }
+
+            frame[tY][tX] = charTetromino;
+        }
+    }
+
+    //Save Frame 
+    if (collision && LastFrame != null) frame = LastFrame;
+    LastFrame = (char[][])frame.Clone();
+    for (int y = 0; y < LastFrame.Length; y++)
+    {
+        LastFrame[y] = (char[])frame[y].Clone();
+    }
+
+    //Draw Preview
+    for (int yField = PLAYFIELD.Length - shapeScope.Length - BORDER; yField >= 0; yField -= 2)
+    {
+        if (CollisionPreview(yField, yScope, shapeScope)) continue;
+
+        for (int y = 0; y < shapeScope.Length && !collision; y++)
+        {
+            for (int x = 0; x < shapeScope[y].Length; x++)
+            {
+                int tY = yField + y;
+
+                if (yScope + shapeScope.Length > tY) continue;
+
+                int tX = TETROMINO.X + x;
+                char charToReplace = PLAYFIELD[tY][tX];
+                char charTetromino = shapeScope[y][x];
+
+                if (charTetromino == ' ') continue;
+
+                if (charToReplace != ' ')
+                {
+                    collision = true;
+                    break;
+                }
+
+                frame[tY][tX] = '•';
+            }
+        }
+
+        break;
+    }
+
+    //Next Square
+    for (int y = 0; y < NEXTTETROMINO.Length; y++)
+    {
+        frame[y] = frame[y].Concat(NEXTTETROMINO[y]).ToArray();
+    }
+
+    //Score Square
+    for (int y = 0; y < SCORE.Length; y++)
+    {
+        int sY = NEXTTETROMINO.Length + y;
+        frame[sY] = frame[sY].Concat(SCORE[y]).ToArray();
+    }
+
+    //Draw Next
+    for (int y = 0; y < TETROMINO.Next.Length; y++)
+    {
+        for (int x = 0; x < TETROMINO.Next[y].Length; x++)
+        {
+            int tY = y + BORDER;
+            int tX = PLAYFIELD[y].Length + x + BORDER;
+            char charTetromino = TETROMINO.Next[y][x];
```

Comment:
I encountered a `System.IndexOutOfRangeException` here while testing the code.

Reply:
I made some improvements, hope that helps.